src/main/java/com/cleanroommc/groovyscript/compat/mods/advancedmortars/Mortar.java

```diff
@@ -0,0 +1,153 @@
+
+    public Mortar() {
+        super();
+    }
+
+    public RecipeBuilder recipeBuilder() {
+        return new RecipeBuilder();
+    }
+
+    @Override
+    @GroovyBlacklist
+    public void onReload() {
+        removeScripted().forEach(this::remove);
+        restoreFromBackup().forEach(recipe -> getTypes(recipe).forEach(type -> add(type, recipe)));
+    }
+
+    public void add(List<String> types, ItemStack output, int duration, List<IIngredient> inputs) {
+        add(types, output, duration, ItemStack.EMPTY, 0.0f, inputs);
+    }
+
+    public void add(List<String> types, ItemStack output, int duration, ItemStack secondaryOutput, float secondaryOutputChance, List<IIngredient> inputs) {
+        if (inputs == null || inputs.size() == 0) return;
+        if (inputs.size() > 8) {
+            GroovyLog.msg("Error adding Advanced Mortars recipe")
+                    .add("maximum number of 8 input ingredients exceeded: " + inputs.size())
+                    .error()
+                    .post();
+            return;
+        }
+        for (String type : types) {
+            EnumMortarType enumMortarType = EnumMortarType.fromName(type);
+            if (enumMortarType == null) {
+                GroovyLog.msg("Error adding Advanced Mortars recipe")
+                        .add("invalid mortar type: " + type)
+                        .add("valid types are: " + Arrays.toString(EnumMortarType.NAMES))
+                        .error()
+                        .post();
+                return;
+            }
+            add(enumMortarType, new RecipeMortar(output, duration, secondaryOutput, secondaryOutputChance, IngredientHelper.toIngredientNonNullList(inputs)));
+        }
+    }
+
+    public void add(EnumMortarType type, RecipeMortar recipe) {
+        List<RecipeMortar> list = ((RegistryRecipeMortarAccessor) MortarAPI.RECIPE_REGISTRY).getRecipeMap().computeIfAbsent(type, (k) -> new ArrayList<>());
+        list.add(recipe);
+        addScripted(recipe);
+    }
+
+    public void remove(RecipeMortar recipe) {
+        ((RegistryRecipeMortarAccessor) MortarAPI.RECIPE_REGISTRY).getRecipeMap().values().forEach(list -> list.removeIf(r -> r == recipe));
+        addBackup(recipe);
+    }
+
+    public static List<EnumMortarType> getTypes(RecipeMortar recipe) {
+        return ((RegistryRecipeMortarAccessor) MortarAPI.RECIPE_REGISTRY)
+                .getRecipeMap()
+                .entrySet()
+                .stream()
+                .filter(x -> x.getValue().contains(recipe))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+    }
+
+    public class RecipeBuilder extends AbstractRecipeBuilder<RecipeMortar> {
+
+        private final List<String> types = new ArrayList<>();
+        private int duration;
+        private ItemStack secondaryOutput = ItemStack.EMPTY;
+        private float secondaryOutputChance = 1.0f;
+
+        public RecipeBuilder type(String... type) {
```

Comment:
i think it would useful if there was a method for each type - eg `public RecipeBuilder diamond() { this.types.add("diamond"); return this; }`. not particularly needed, though, if you dont want to bother.

Reply:
I can do that